### PR TITLE
Modify rule S4023: LayC and improve rule description

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
see also https://github.com/SonarSource/sonar-dotnet/issues/7633#issuecomment-1653156163

### CODE REVIEW

File: rules/S4023/csharp/rule.adoc [hunk 1]
```diff
@@ -1,4 +1,132 @@
...
+{
+    // Use reflection to check if the interface is applied to the type
+    var includeFields = o.GetType()
+        .GetInterfaces().Any(i => i == typeof(IIncludeFields));
+    var options = new JsonSerializerOptions()
+    {
+        // Take decisions based on the presents of the marker
+        IncludeFields = includeFields, 
+    };
+}
```
Comment (cristian-ambrosini-sonarsource) [inline rules/S4023/csharp/rule.adoc:30]: That's a good sample, just a thought: I would simplify it a bit and avoid all the JSON serialization configuration, I think it's not a relevant topic for this rule and might add complexity for the reader.

File: rules/S4023/csharp/rule.adoc [hunk 2]
```diff
@@ -1,4 +1,132 @@
...
+
+* Add members to the interface
+* Remove the useless interface
+* Replace the usage as a marker interface with custom attributes
+
+=== Code examples
+
+==== Noncompliant code example
+
+[source,csharp,diff-id=2,diff-type=noncompliant]
```
Comment (cristian-ambrosini-sonarsource) [inline rules/S4023/csharp/rule.adoc:104]: It's a bit confusing to have one solution in the "Why section" and another solution in the "How to fix it".
I would keep both examples in this section, each under their appropriate bullet point.
- Add members to the interface:
[Noncompliant/Compliant snippet]
- Replace the usage as a marker interface with custom attributes
[Noncompliant/Compliant snippet]

**Review by cristian-ambrosini-sonarsource — CHANGES_REQUESTED** [review]
Left a few suggestions

File: rules/S4023/csharp/rule.adoc (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply rules/S4023/csharp/rule.adoc:30]: If I remove that part, the `includeFields` variable is mysterious to the reader: What is this used for? By assigning it to `JsonSerializerOptions.IncludeFields` it is obvious to the reader and he can look it up to learn more about it. I think it is essential to the understanding of the `Serialize` method purpose.

File: rules/S4023/csharp/rule.adoc [hunk 3]
```diff
@@ -1,4 +1,138 @@
-include::../rule.adoc[]
+Interfaces describe functionality that an implementing class has to provide. An empty interface doesn't add any such functional requirements and is as such useless.
```
Comment (cristian-ambrosini-sonarsource) [inline rules/S4023/csharp/rule.adoc:1]:
```suggestion
Empty interfaces should be avoided as they do not provide any functional requirements for implementing classes.
```

File: rules/S4023/csharp/rule.adoc [hunk 4]
```diff
@@ -1,4 +1,138 @@
...
+
+=== Exceptions
+
+This rule doesn't raise in any of the following cases:
+
+==== Aggregation of multiple interfaces
+
+[source,csharp]
+----
+public interface IAggregate: IComparable, IFormattable { } // Compliant
```
Comment (cristian-ambrosini-sonarsource) [inline rules/S4023/csharp/rule.adoc:15]: The first snippet has a `// Compliant` comment, the second an extended `// Compliant: explanation` and the third none. IMO we can improve a bit the consistency

**Review by cristian-ambrosini-sonarsource — CHANGES_REQUESTED** [review]

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]
LGTM!

File: rules/S4023/csharp/rule.adoc [hunk 5]
```diff
@@ -1,4 +1,138 @@
...
+Serialize(new OptInToIncludeFields());
+
+void Serialize<T>(T o)
+{
+    // Use reflection to check if the interface is applied to the type
+    var includeFields = o.GetType()
+        .GetInterfaces().Any(i => i == typeof(IIncludeFields));
+    var options = new JsonSerializerOptions()
+    {
+        // Take decisions based on the presents of the marker
```
Comment (bart-vmware) [inline rules/S4023/csharp/rule.adoc:96]: ~presents~ presence

File: rules/S4023/csharp/rule.adoc [hunk 6]
```diff
@@ -1,4 +1,138 @@
...
+Serialize(new OptInToIncludeFields());
+
+void Serialize<T>(T o)
+{
+    // Use reflection to check if the attribute is applied to the type
+    var includeFields = o.GetType()
+        .GetCustomAttributes(typeof(IncludeFieldsAttribute), inherit: true).Any();
+    var options = new JsonSerializerOptions()
+    {
+        // Take decisions based on the presents of the marker
```
Comment (bart-vmware) [inline rules/S4023/csharp/rule.adoc:122]: ~presents~ presence

Comment (martin-strecker-sonarsource) [reply rules/S4023/csharp/rule.adoc:122]: Thank you! #2755

**Pull request merged**